apollo-backend/src/main/java/io/logz/apollo/status/ServiceStatusHandler.java

```diff
@@ -0,0 +1,133 @@
+                isServiceUndeployed(serviceId, environmentId, Optional.of(group), timeUnit, maxUndeployedTime)).collect(Collectors.toList());
+    }
+
+    private boolean isServiceUndeployed(int serviceId, int environmentId, Optional<Group> group, TimeUnit timeUnit, int maxUndeployedTime) {
+        Optional<Date> latestDeploymentDate;
+        Optional<Date> latestUpdatedDate = getLatestDeployableVersionDateByService(serviceId);
+        if(group.isPresent()) {
+            latestDeploymentDate = getLatestDeploymentDateByServiceAndGroupAndEnvironment(serviceId, environmentId, group.get().getName());
+        }
+        else {
+            latestDeploymentDate = getLatestDeploymentDateByServiceAndEnvironment(serviceId, environmentId);
+        }
+        return isDeploymentDateExpired(latestDeploymentDate, latestUpdatedDate, timeUnit, maxUndeployedTime);
+    }
+
+    private boolean isDeploymentDateExpired(Optional<Date> latestDeploymentDate, Optional<Date> latestUpdatedDate, TimeUnit timeUnit, int maxUndeployedTime) {
+        if (latestDeploymentDate.isPresent() && latestUpdatedDate.isPresent()) {
+            return getTimeDiff(latestDeploymentDate.get(), latestUpdatedDate.get(), timeUnit) > maxUndeployedTime;
+        }
+        if (latestUpdatedDate.isPresent()) {
+            return latestUpdatedDate.get().compareTo(calculateMaxDateSinceLatestUpdate(timeUnit, maxUndeployedTime)) > 0;
+        }
+        return false;
+    }
+
+    private Optional<Date> getLatestDeploymentDateByServiceAndEnvironment(int serviceId, int environmentId) {
+        Optional<Deployment> deployment = Optional.ofNullable(deploymentDao.getLatestDeploymentOfServiceAndEnvironment(serviceId, environmentId));
+        return deployment.isPresent() ? Optional.of(deployment.get().getLastUpdate()) : Optional.empty();
+    }
+
+    private Optional<Date> getLatestDeploymentDateByServiceAndGroupAndEnvironment(int serviceId, int environmentId, String groupName) {
+        Optional<Deployment> deployment = Optional.ofNullable(deploymentDao.getLatestDeploymentOfServiceAndEnvironmentByGroupName(serviceId, environmentId, groupName));
+        return deployment.isPresent() ? Optional.of(deployment.get().getLastUpdate()) : Optional.empty();
+    }
+
+    private Optional<Date> getLatestDeployableVersionDateByService(int serviceId) {
+        Optional<DeployableVersion> latestDeployableVersion = Optional.ofNullable(deployableVersionDao.getLatestDeployableVersionByServiceId(serviceId));
+        return latestDeployableVersion.isPresent() ? Optional.of(latestDeployableVersion.get().getCommitDate()) : Optional.empty();
+    }
+
+    private long getTimeDiff(Date date1, Date date2, TimeUnit resultTimeUnit) {
+        return resultTimeUnit.convert(Duration.between(date1.toInstant(), date2.toInstant()).toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    private Date calculateMaxDateSinceLatestUpdate(TimeUnit timeUnit, int maxTimeInterval) {
```

Comment:
I dont understand this function

Reply:
to get the last possible date Since the last deployable version